### PR TITLE
Add basic authentication using HMAC for federates joining a federation

### PR DESCRIPTION
**Pull request by hokeun**
This PR is a submodule PR of the PR lf-lang/lingua-franca#1432 and is the first step towards resolving lf-lang/lingua-franca#1146. Both PRs should be merged to work.

Overall information is also in lf-lang/lingua-franca#1432. This PR details the HMAC authentication implementation.

This PR is about the RTI.c and federate.c authentication implementation.

The main idea of security is not to send the federation ID in plain text, by using it as an HMAC key.

### 1. RTI_HELLO (RTI -> Federate) ###
The RTI sends a RTI_HELLO message to the federate when TCP connection is established. The message is in this order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- Eight bytes for randomly created RTI's nonce.
### 2. FED_RESPONSE  (Federate -> RTI) ###
The federate sends a response message to the RTI. It creates it's own nonce, and makes a HMAC tag using the federate ID as the key. The message is in the following order:
- One byte equal to MSG_TYPE_FED_RESPONSE.
- Eight bytes for federate's nonce.
- Two bytes for the federate ID.
- 32 bytes for HMAC tag based on SHA256. 

 The HMAC tag is composed of the following order:
- One byte equal to MSG_TYPE_FED_RESPONSE.
- Two bytes for the federate ID.
- Eight bytes for received RTI's nonce.

### 3. RTI_RESPONSE  (RTI -> Federate) ###
The RTI sends a response message to the federate. It checks the received HMAC created by the federate, and also makes a HMAC tag. The message is in the following order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- 32 bytes for HMAC tag based on SHA256.

The HMAC tag is composed of the following order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- Eight bytes for received federate's nonce.

### Federate Check ###
The federate finally checks the received HMAC created by the RTI.

## Discussion ##
The 3-way handshake protocol does not change the existing federation id check process. ```receive_and_check_fed_id_message()``` still always happen after the 3-way handshake. So, the ```auth:true``` option only enables the 3-way handshake. Changing the protocol will be discussed at the future.

Also, the ```auth:true``` option will be changed from boolean to string in future.

### CODE REVIEW

File: core/federated/federate.c [hunk 1]
```diff
@@ -856,6 +856,24 @@ void connect_to_federate(uint16_t remote_federate_id) {
     }
 }
 
+void perform_hmac_authentication(int rti_socket) {
```
Comment (edwardalee) [inline core/federated/federate.c:859]: Method needs a comment.
It seems all it does is print the message received from the RTI?
The method name implies much more.

File: core/federated/RTI/rti.c [hunk 2]
```diff
@@ -1790,6 +1791,18 @@ int receive_udp_message_and_set_up_clock_sync(int socket_id, uint16_t fed_id) {
     return 1;
 }
 
+/**
+ * Send RTI hello message for authenticated joining of federation.
+ * 
+ * @param socket Socket for the incoming federate tryting to authenticate.
+ */
+void send_rti_hello(int socket) {
```
Comment (edwardalee) [inline core/federated/RTI/rti.c:1799]: Unclear how this does any authentication.
Seems to just send random bytes.
Comment should explain what a "hello message" is.

**Review by edwardalee — COMMENTED** [review]
I realize this is a draft, but it doesn't seem to perform any authentication. Guess this isn't ready for review?

**Comment by hokeun** [comment]
> I realize this is a draft, but it doesn't seem to perform any authentication. Guess this isn't ready for review?

Thanks, @edwardalee for checking! You're right, this is not ready for review yet. I created this draft so that I can easily share my progress with others using `Files Changed` tap. I'll surely let you know when this is ready for review!

**Comment by Jakio815** [comment]
This PR is a submodule PR of the PR lf-lang/lingua-franca#1432 and resolves lf-lang/lingua-franca#1146. Both PRs should be merged to work.

Please read lf-lang/lingua-franca#1432 before this comment. This PR is the details of the HMAC authentication.

This PR is about the RTI.c and federate.c authentication implementation.

The main idea of security is to hide the federation ID, by using it as a HMAC key.

## 3-Way Handshake ##
### RTI_HELLO (RTI -> Federate) ###
The RTI sends a RTI_HELLO message to the federate when TCP connection is established. The message is in this order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- Eight bytes for randomly created RTI's nonce.
### FED_RESPONSE  (Federate -> RTI) ###
The federate sends a response message to the RTI. It creates it's own nonce, and makes a HMAC tag using the federate ID as the key. The message is in the following order:
- One byte equal to MSG_TYPE_FED_RESPONSE.
- Eight bytes for federate's nonce.
- Two bytes for the federate ID.
- 32 bytes for HMAC tag based on SHA256. 

 The HMAC tag is composed of the following order:
- One byte equal to MSG_TYPE_FED_RESPONSE.
- Two bytes for the federate ID.
- Eight bytes for received RTI's nonce.

### RTI_RESPONSE  (RTI -> Federate) ###
The RTI sends a response message to the federate. It checks the received HMAC created by the federate, and also makes a HMAC tag. The message is in the following order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- 32 bytes for HMAC tag based on SHA256.

The HMAC tag is composed of the following order:
- One byte equal to MSG_TYPE_RTI_RESPONSE.
- Eight bytes for received federate's nonce.

### Federate Check ###
The federate finally checks the received HMAC created by the RTI.

File: core/federated/federate.c (hunk 1 above)
Comment (Jakio815) [reply core/federated/federate.c:873]: Thanks for reviewing. Now this PR is ready for review. As @hokeun  commented below, it was a draft PR before.

File: core/federated/RTI/rti.c (hunk 2 above)
Comment (Jakio815) [reply core/federated/RTI/rti.c]: As same as above, all functions are ready with comments. Thank you for reviewing.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -26,30 +26,30 @@ jobs:
...
     with:
       runtime-ref: ${{ github.ref }}
       compiler-ref: ${{ needs.fetch-lf.outputs.ref }}
       scheduler: GEDF_NP_CI
 
   lf-adaptive:
     needs: fetch-lf
-    uses: lf-lang/lingua-franca/.github/workflows/c-tests.yml@master
+    uses: lf-lang/lingua-franca/.github/workflows/c-tests.yml@auth
     with:
```
Comment (Jakio815) [inline .github/workflows/ci.yml:53]: These changes are only for CI tests.

File: include/core/federated/net_common.h [hunk 4]
```diff
@@ -330,6 +330,53 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 
+/////////// Messages used for authenticated federation. ///////////////
+/**
+ * Byte identifying a message from an RTI to a federate containing
+ * RTI's 8-byte random nonce for HMAC-based authentication. The RTI sends this
+ * message to an incoming federate when TCP connection is established
+ * between the RTI and the federate.
+ * The next eight bytes are RTI's 8-byte nonce (RTI nonce).
+ */
+#define MSG_TYPE_RTI_HELLO 100
```
Comment (lhstrh) [inline include/core/federated/net_common.h:341]: Why not call it `MSG_TYPE_RTI_NONCE`?

**Review by lhstrh — APPROVED** [review]
This looks good to me. Though I think that we might want to enable security by default and provide the option to disable it for whatever reason...

File: include/core/federated/net_common.h (hunk 4 above)
Comment (Jakio815) [reply include/core/federated/net_common.h]: Fixed this part.

**Pull request merged**